### PR TITLE
Changes to array navigation

### PR DESCRIPTION
**Pull request by pavlopoul**

### CODE REVIEW

File: vxquery-core/src/main/java/org/apache/vxquery/runtime/functions/json/ValueScalarEvaluator.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+        try {
+            ab.reset(mvs);
+            TaggedValuePointable tvp1 = args[0];
+            if (tvp1.getTag() == ValueTag.ARRAY_TAG) {
+                tvp1.getValue(ap);
+                TaggedValuePointable tvp2 = args[1];
+                TaggedValuePointable tempTvp = ppool.takeOne(TaggedValuePointable.class);
+                tvp2.getValue(lp);
+                if ((int) lp.getLong() > ap.getEntryCount())
+                    throw new RuntimeException("The argument is out of bounds");
```
Comment (westmann) [inline vxquery-core/src/main/java/org/apache/vxquery/runtime/functions/json/ValueScalarEvaluator.java]: Shouldn't we return an empty sequence here?

Comment (westmann) [inline vxquery-core/src/main/java/org/apache/vxquery/runtime/functions/json/ValueScalarEvaluator.java]: Put braces around the single line?

File: vxquery-core/src/main/java/org/apache/vxquery/runtime/functions/json/ValueScalarEvaluator.java [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+                ap.getEntry((int) lp.getLong() - 1, tempTvp);
+                try {
+                    if (ap.getEntryCount() != 0)
```
Comment (westmann) [inline vxquery-core/src/main/java/org/apache/vxquery/runtime/functions/json/ValueScalarEvaluator.java]: How about

```
ab.addItem(ap.getEntryCount() != 0 ? tempTvp : tvp1);
```

instead?

Also, it seems that there's too much distance between the allocation of the tempTvp and the start of the try block. What happens if an exception gets thrown between them.

File: vxquery-core/src/main/java/org/apache/vxquery/runtime/functions/json/ValueScalarEvaluator.java [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+                        ab.addItem(tempTvp);
+                    else
+                        ab.addItem(tvp1);
+                } finally {
+                    ppool.giveBack(tempTvp);
+                }
+                ab.finish();
+            }
+        } catch (IOException e) {
+            e.printStackTrace();
```
Comment (westmann) [inline vxquery-core/src/main/java/org/apache/vxquery/runtime/functions/json/ValueScalarEvaluator.java]: Please don't swallow exceptions. It seems that we won't get the correct result here and so we should abort. One option would be to wrap the exception into a SystemException with error code SYSE0001.

File: vxquery-core/src/main/java/org/apache/vxquery/runtime/functions/misc/FnSizeScalarEvaluatorFactory.java [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+        @Override
+        protected void evaluate(TaggedValuePointable[] args, IPointable result) throws SystemException {
+            abvs.reset();
+            TaggedValuePointable tvp = args[0];
+            if (tvp.getTag() == ValueTag.ARRAY_TAG) {
+                tvp.getValue(ap);
+                DataOutput out = abvs.getDataOutput();
+                ap.getEntryCount();
+                try {
+                    out.write(ValueTag.XS_INT_TAG);
```
Comment (westmann) [inline vxquery-core/src/main/java/org/apache/vxquery/runtime/functions/misc/FnSizeScalarEvaluatorFactory.java]: Shouldn't this be an XS_INTEGER?

File: vxquery-core/src/main/java/org/apache/vxquery/functions/builtin-operators.xml [hunk 5]
```diff
@@ -646,6 +646,14 @@
         <return type="xs:anyAtomicType?"/>
         <runtime type="scalar" class="org.apache.vxquery.runtime.functions.arithmetic.AddScalarEvaluatorFactory"/>
     </operator>
+    
+    <!-- opext:value($expr as xs:item(), $arg as xs:anyAtomicType?) as xs:anyAtomicType? -->
+    <operator name="opext:value">
+        <param name="expr" type="item()"/>
+        <param name="argExpr" type="xs:anyAtomicType"/>
+        <return type="xs:anyAtomicType"/>
```
Comment (westmann) [inline vxquery-core/src/main/java/org/apache/vxquery/functions/builtin-operators.xml]: Why does the result type need to be atomic?

File: vxquery-core/src/main/java/org/apache/vxquery/serializer/XMLSerializer.java [hunk 6]
```diff
@@ -546,6 +527,26 @@ private void printSequence(PrintStream ps, SequencePointable seqp, String betwee
...
+                ap.getEntry(i, tvp);
+                print(tvp.getByteArray(), tvp.getStartOffset(), tvp.getLength(), ps);
+                if (i != len - 1) {
+                    ps.append(',');
+                }
+            }
+            ps.append(']');
+        } finally {
+            pp.giveBack(ap);
+            pp.giveBack(tvp);
```
Comment (westmann) [inline vxquery-core/src/main/java/org/apache/vxquery/serializer/XMLSerializer.java]: Could we give the tvp back in the same scope that it is allocated in?

File: vxquery-core/src/main/java/org/apache/vxquery/xmlquery/ast/ASTTag.java [hunk 7]
```diff
@@ -80,6 +81,7 @@
     AXIS_STEP,
     NAME_TEST,
     FILTER_EXPRESSION,
```
Comment (westmann) [inline vxquery-core/src/main/java/org/apache/vxquery/xmlquery/ast/ASTTag.java]: Do we still need the FILTER_EXPRESSION or can that we expressed by the POSTFIX_EXPRESSION as well?

File: vxquery-core/src/main/javacc/xquery-grammar.jj [hunk 8]
```diff
@@ -1541,7 +1541,8 @@ ASTNode StepExpr():
     ASTNode expr;
 }
 {
-    LOOKAHEAD(3) expr = FilterExpr() {
+    
+    LOOKAHEAD(3)expr = PostfixExpr() {
```
Comment (westmann) [inline vxquery-core/src/main/javacc/xquery-grammar.jj]: space?

File: vxquery-core/src/main/javacc/xquery-grammar.jj [hunk 9]
```diff
@@ -1743,17 +1744,54 @@ NameTestNode Wildcard():
...
+        PostfixExprNode pe = new PostfixExprNode(expr.getSourceLocation());
+        pe.setExpr(expr);
+        pe.setArgs(argList);  
+        return pe;
+    }
+}
+
+ASTNode Arguments()    :
+{
+    List<ASTNode> expr = new ArrayList<ASTNode>();
```
Comment (westmann) [inline vxquery-core/src/main/javacc/xquery-grammar.jj]: Why do we need to allocate here?

Comment (westmann) [inline vxquery-core/src/main/javacc/xquery-grammar.jj]: Also we might want to have a more descriptive variable name.

File: vxquery-xtest/src/test/resources/Queries/XQuery/Json/Array/Navigation/q02_array_navigation.xq [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+   software distributed under the License is distributed on an
+   "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+   KIND, either express or implied.  See the License for the
+   specific language governing permissions and limitations
+   under the License. :)
+
+(: Json Array Navigation Query :)
+(: Just navigate through an array :)
+    let $x:=[1,[2]]
+    return $x(2)(1)
```
Comment (westmann) [inline vxquery-xtest/src/test/resources/Queries/XQuery/Json/Array/Navigation/q02_array_navigation.xq]: Add a new line at the end?

File: vxquery-xtest/src/test/resources/Queries/XQuery/Json/Array/Navigation/q03_array_navigation.xq [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+   software distributed under the License is distributed on an
+   "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+   KIND, either express or implied.  See the License for the
+   specific language governing permissions and limitations
+   under the License. :)
+
+(: Json Array Size Query :)
+(: Size of an array :)
+    let $x:=[2,[2,1,3]]
+    return size($x)
```
Comment (westmann) [inline vxquery-xtest/src/test/resources/Queries/XQuery/Json/Array/Navigation/q03_array_navigation.xq]: Add a new line at the end?

File: vxquery-xtest/src/test/resources/cat/JsonArrayNavigationQueries.xml [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+      <description>Navigate through an empty array.</description>
+      <query name="q04_array_navigation" date="2016-06-18"/>
+      <output-file compare="Text">q04_array_navigation.txt</output-file>
+   </test-case>
+   <test-case name="json-array-navigation-q05" FilePath="Json/Array/Navigation" Creator="Christina Pavlopoulou">
+      <description>Ask for an index out of the array size.</description>
+      <query name="q04_array_navigation" date="2016-06-18"/>
+      <output-file compare="Text">q05_array_navigation.txt</output-file>
+   </test-case>
+</test-group>
```
Comment (westmann) [inline vxquery-xtest/src/test/resources/cat/JsonArrayNavigationQueries.xml]: Add a new line at the end?

File: vxquery-xtest/src/test/resources/ExpectedTestResults/Json/Array/q02_array.txt [hunk 13]
```diff
@@ -1 +1 @@
-[1,"string",3.1]
\ No newline at end of file
+[1,"string",3.1]
```
Comment (westmann) [inline vxquery-xtest/src/test/resources/ExpectedTestResults/Json/Array/q02_array.txt:1]: Do we have a query file for this expected result file?

File: vxquery-xtest/src/test/resources/cat/JsonArrayNavigationQueries.xml [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+      <output-file compare="Text">q03_array_navigation.txt</output-file>
+   </test-case>
+   <test-case name="json-array-navigation-q04" FilePath="Json/Array/Navigation" Creator="Christina Pavlopoulou">
+      <description>Navigate through an empty array.</description>
+      <query name="q04_array_navigation" date="2016-06-18"/>
+      <output-file compare="Text">q04_array_navigation.txt</output-file>
+   </test-case>
+   <test-case name="json-array-navigation-q05" FilePath="Json/Array/Navigation" Creator="Christina Pavlopoulou">
+      <description>Ask for an index out of the array size.</description>
+      <query name="q04_array_navigation" date="2016-06-18"/>
```
Comment (westmann) [inline vxquery-xtest/src/test/resources/cat/JsonArrayNavigationQueries.xml]: Should this be `q05`?

File: vxquery-core/src/main/java/org/apache/vxquery/xmlquery/translator/XMLQueryTranslator.java [hunk 15]
```diff
@@ -1540,6 +1542,38 @@ private LogicalVariable translatePathExpr(PathExprNode pe, TranslationContext tC
...
+                        for (ASTNode an : args) {
+                            if (an.getTag() == ASTTag.ARGUMENT_LIST) {
+                                ArgumentListNode argNode = (ArgumentListNode) an;
+                                for (ASTNode en : argNode.getExpression()) {
+                                    ILogicalExpression argument = vre(translateExpression(en, tCtx));
+                                    arguments.add(argument);
+                                    if (args.size() == 1) {
+                                        ctxExpr = sfce(BuiltinOperators.VALUE, expr, argument);
+                                    } else {
+                                        if (an == args.get(0))
```
Comment (westmann) [inline vxquery-core/src/main/java/org/apache/vxquery/xmlquery/translator/XMLQueryTranslator.java]: I'm not sure that I understand the position logic here?
Could you add a comment that explains?

File: vxquery-core/src/main/java/org/apache/vxquery/runtime/functions/json/ValueScalarEvaluator.java [hunk 16]
```diff
@@ -0,0 +1,82 @@
...
+        this.ctx = ctx;
+        ap = (ArrayPointable) ArrayPointable.FACTORY.createPointable();
+        lp = (LongPointable) LongPointable.FACTORY.createPointable();
+        mvs = new ArrayBackedValueStorage();
+        ab = new SequenceBuilder();
+    }
+
+    @Override
+    protected void evaluate(TaggedValuePointable[] args, IPointable result) throws SystemException {
+        TaggedValuePointable tvp1;
```
Comment (prestoncarman) [inline vxquery-core/src/main/java/org/apache/vxquery/runtime/functions/json/ValueScalarEvaluator.java]: If check if the arguments are correct. Then process the result. See other evaluators for examples.

File: vxquery-core/src/main/java/org/apache/vxquery/runtime/functions/misc/FnSizeScalarEvaluatorFactory.java [hunk 17]
```diff
@@ -0,0 +1,76 @@
...
+        final ArrayBackedValueStorage abvs = new ArrayBackedValueStorage();
+        final ArrayPointable ap = (ArrayPointable) ArrayPointable.FACTORY.createPointable();
+
+        public FnSizeScalarEvaluator(IScalarEvaluator[] args) {
+            super(args);
+        }
+
+        @Override
+        protected void evaluate(TaggedValuePointable[] args, IPointable result) throws SystemException {
+            abvs.reset();
```
Comment (prestoncarman) [inline vxquery-core/src/main/java/org/apache/vxquery/runtime/functions/misc/FnSizeScalarEvaluatorFactory.java]: Check arguments before processing result. If the arguments are incorrect, throw an error.

File: vxquery-core/src/main/java/org/apache/vxquery/serializer/XMLSerializer.java [hunk 18]
```diff
@@ -546,6 +527,25 @@ private void printSequence(PrintStream ps, SequencePointable seqp, String betwee
         }
     }
 
+    private void printArray(PrintStream ps, TaggedValuePointable tvp) {
```
Comment (prestoncarman) [inline vxquery-core/src/main/java/org/apache/vxquery/serializer/XMLSerializer.java]: I don't think you need change this file.

**Pull request closed without merging**